alf/utils/schedulers.py

```diff
@@ -194,6 +194,95 @@ def __repr__(self):
 
 
+@alf.configurable
+class CyclicalScheduler(Scheduler):
+    """The cyclical scheduler where the value changes cyclically between two bounds.
+    Reference:
+    ::
+        Leslie N. Smith Cyclical Learning Rates for Training Neural Networks, 2017
+        (https://arxiv.org/pdf/1506.01186.pdf)
+
+    This implementation generalizes the original methods in two ways: 1) the
+    initial value can start from either the lower-bound (as in the original method),
+    or upper bound; 2) apart from the linear switching between the bounds, we
+    also support step mode of switching.
+
+    In terms of applications, beyond the standard case of using a cyclical
+    learning rate to improve the learning behavior during NN training, this
+    scheduler is also useful in other cases. One example is in reinforcement
+    learning, sometimes we want to update the parameters of different modules at
+    difference paces. For example, in TD3, we want to update the policy every other
+    updates. In this case, we can use a ``CyclicalScheduler`` with ``step``
+    switching mode to achieve this. Similar cases also appears in Dreamer.
+    """
+
+    def __init__(self,
+                 progress_type,
+                 base_lr,
+                 bound_lr,
+                 half_cycle_size,
+                 switch_mode='step'):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+            base_lr (float): the base learning rate, representing the starting
+                value.
+            bound_lr (float): the value of the learning rate on the other bound.
+                The value of ``bound_lr`` could be either larger or smaller than
+                the value of ``base_lr``.
+            half_cycle_size (int|float): the length of half a cycle. Its actual
+                length is based on the ``progress_type``. For example, if in
+                "iterations" mode, it means the lr value will reach the opposite
+                bound every ``half_cycle_size`` iterations.
+            switch_mode (str): the way to switch from one bound to the other.
+                Currently support the following modes:
+                - step: directly jump from one mode to the other every half cycle
+                - linear: linearly move from one mode to the other every half cycle
+        """
+        super().__init__(progress_type)
+        self._base_lr = base_lr
+        self._bound_lr = bound_lr
+        self._half_cycle_size = half_cycle_size
+        self._cycle_size = half_cycle_size * 2
+        assert switch_mode in {
+            "step", "linear"
+        }, ("unsupportted switch mode {}".format(switch_mode))
+        self._switch_mode = switch_mode
+        self._current_value = base_lr
+
+    def __call__(self):
+        progress = self.progress()
+
+        # round to one decimal place to facilitate the stage comparison below
+        progress_in_half_cycle = round(
```

Comment:
It seems that rounding is not necessary.

Reply:
```round``` is used to handle some cases that could occur  due to numerical reasons when ```progress_type``` is ```percent```.

For example, for a case where ```half_cycle_size=0.1```, when the ```progress=0.6```, both ```progress_in_half_cycle``` and ```progress_in_cycle``` should be 0.  However, due to numerical reasons,  the following results are obtained:

<img width="322" alt="image" src="https://user-images.githubusercontent.com/21375027/184963488-0e5deee7-c186-45ea-9f9b-dbf34f7ccc42.png">

Rounding is used to avoid this issue.  For other ``progress_type```s, it is unnecessary and introduces an approximation.

Now changed to only apply it when ```progress_type``` is ```percent```.